File: src/test/java/dev/kensa/matcher/IndexMatcher.java
package dev.kensa.matcher;

import dev.kensa.sentence.Token;
import dev.kensa.sentence.scanner.Index;
import org.assertj.core.api.Condition;

import static dev.kensa.matcher.Conditions.equalTo;
import static org.assertj.core.api.Assertions.allOf;

public final class IndexMatcher {

    public static Condition<Index> theSameAs(Index expected) {
        return allOf(
                aTypeOf(expected.type()),
                aStartOf(expected.start()),
                anEndOf(expected.end())
        );
    }

    public static Condition<Index> aTypeOf(Token.Type expected) {
        return equalTo("Type", Index::type, expected);
    }

    public static Condition<Index> aStartOf(int expected) {
        return equalTo("Start Index", Index::start, expected);
    }

    public static Condition<Index> anEndOf(int expected) {
        return equalTo("End Index", Index::end, expected);
    }
}

File: src/test/java/dev/kensa/matcher/IndexMatcher.kt
package dev.kensa.matcher

import dev.kensa.matcher.Conditions.equalTo
import dev.kensa.sentence.TokenType
import dev.kensa.sentence.scanner.Index
import org.assertj.core.api.Assertions
import org.assertj.core.api.Condition

object IndexMatcher {
    fun theSameAs(expected: Index): Condition<Index> =
            Assertions.allOf(
                    aTypeOf(expected.type),
                    aStartOf(expected.start),
                    anEndOf(expected.end)
            )

    fun aTypeOf(expected: TokenType) = equalTo("Type", Index::type, expected)

    fun aStartOf(expected: Int) = equalTo("Start Index", Index::start, expected)

    fun anEndOf(expected: Int) = equalTo("End Index", Index::end, expected)
}